src/app/CommandHandler.cpp

```diff
@@ -65,25 +65,24 @@ CHIP_ERROR CommandHandler::AllocateBuffer()
 CHIP_ERROR CommandHandler::OnInvokeCommandRequest(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
                                                   System::PacketBufferHandle && payload)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
     System::PacketBufferHandle response;
-
     VerifyOrReturnError(mState == CommandState::Idle, CHIP_ERROR_INCORRECT_STATE);
 
     // NOTE: we already know this is an InvokeCommand Request message because we explicitly registered with the
     // Exchange Manager for unsolicited InvokeCommand Requests.
-
     mpExchangeCtx = ec;
 
-    err = ProcessInvokeRequest(std::move(payload));
-    SuccessOrExit(err);
-
-    err = SendCommandResponse();
-    SuccessOrExit(err);
+    // Use the RAII feature, if this is the only Handle when this function returns, DecRef will trigger sending response.
+    Handle workHandle(this);
+    ReturnErrorOnFailure(ProcessInvokeRequest(std::move(payload)));
```

Comment:
So what happens if `ProcessInvokeRequest()` returns error _but_ in the process someone has taken a handle to us?  For example, we had multiple commands in the list, the first started doing async work, the second failed to parse and we bailed out of the loop in `ProcessInvokeRequest` with an error.

We will not close ourselves, but the exchange _will_ close, and we will have a dangling pointer to it.

It seems to me that we need to either do the refcount check even if ProcessInvokeRequest fails, or more simply unconditionally call `WillSendMessage` before we call `ProcessInvokeRequest`, since we will in fact do the send once our refcount goes to 0.

Reply:
I'd prefer the unconditional call, since we know for certain (till we implement SuppressResponse and groupcast commands) that we're always going to send out a response.

Reply:
Updated to unconditional call.